Optimization: Shared allocator without mutex

kython28 opened the pull request:
Hello again :D

After my yesterday PR and your answer about finding a solution without using a mutex, I wanted to try if maybe... there is a way. And there is. In this new implementation of the Shared Allocator, we can get an existing arena without using a mutex. The mutex is being used only when we are creating a batch of arenas.

All the test cases have passed, and I also added 2 more. And the best thing is... the benchmarking test is showing that the Shared Allocator outperforms the Global Allocator (at least for mixed, small, and medium sizes).

I wrote this on another branch because maybe you want to take more time looking at these changes, meanwhile the others are more simple.

kython28 committed:
chore: preparing code for compiling with zig 0.14.0

kython28 committed:
fix: Fixing up data race detected in shared_allocator

Ref: https://github.com/joadnacer/jdz_allocator/issues/9

kython28 committed:
chore: Updating `bench.zig` for zig 0.14.0

kython28 committed:
build: Updating version to 0.1.0

kython28 committed:
refactor: Trying to reduce mutexes usage

kython28 committed:
chore: ...

kython28 committed:
refactor: Improving performance for shared_allocator reducing mutexes usage

kython28 committed:
chore: Just modifying this to detect errors while running tests

kython28 committed:
tests: Adding more test cases for shared allocator

kython28 committed:
fix: Solving problem when the gpa is created multiple times

kython28 committed:
test: Adding more tests cases

kython28 committed:
build: Adding sanitize-thread flag to shared-allocator tests

kython28 committed:
chore: ...

kython28 committed:
fix: Fixing up data race

kython28 commented:
I identified some problems and got some ideas. I will reopen this when i finish 😀

I also detected data race in the global allocator when calling `deInitThread`

kython28 committed:
chore: Replacing assert for custom one

I think assert is good when you in debugging mode and maybe for
ReleaseSafe. With this change we can remove it detecting release mode

kython28 committed:
Feat: Adding arenas recycling flow

In my previous changes i didn't take into account several situations and
scenarios. This is something that already exists in the previous version
but missing in this one.

kython28 committed:
Refactor: Improving a bit performance for global allocator

Just replacing the previous version for a LIFO

joadnacer commented:
great - do you maybe want to remove the mutex commit from your previous pr and I can merge just the 0.14.0 part?

kython28 committed:
refactor: Simplify SharedArenaHandler slot management and remove unused state tracking

kython28 committed:
test: Add multi-allocator parallel allocation tests with different configurations

kython28 committed:
test: Add synchronized memory allocation and deallocation test

kython28 committed:
refactor: Reorder Span struct fields for better memory alignment

kython28 committed:
refactor: Improve atomic memory ordering in MPMC queue operations

kython28 committed:
feat: Optimize shared arena batch size indexing using bitwise AND

kython28 committed:
refactor: Improve type safety and memory ordering in concurrent data structures

kython28 committed:
refactor: Disable parallel allocation test and adjust shared arena batch size

kython28 committed:
refactor: Optimize arena index calculation using bitwise AND

kython28 commented:
Hello, 

I have finished, and everything appears to be working well so far. I haven't resolved the data race issue in the global allocator yet, but maybe I'll address it in a future pull request. I've left a commented test in the file that can help you to replicate the problem. 

Changes applied: 

Currently, we have a mechanism that works similarly to the previous approach: 

When a new thread needs to allocate memory, it looks for an arena and attempts to adopt it. Previously, this was a linear search through arenas to find an available one. Now, there's a more direct method to obtain and find a free arena. This new approach doesn't require a mutex like the previous one, using only atomic operations to find an available arena. In the best-case scenario, a thread can directly access an arena it previously used without atomic operations. The mutex is only used when creating a new batch of arenas, so it's only needed the first time (as well for atomic operations). 

There are some additional modifications and bug fixes.

kython28 committed:
chore: Bump version to 0.2.0

kython28 committed:
refactor: Add inline tryLock method to DummyMutex struct

kython28 committed:
chore: Bump version to 0.2.1

kython28 committed:
fix: Update DummyMutex tryLock method to return bool

kython28 committed:
fix: Implement tryLock method in DummyMutex to always return true

joadnacer commented on src/bounded_mpmc_queue.zig:
```diff
@@ -52,12 +52,12 @@ pub fn BoundedMpmcQueue(comptime T: type, comptime buffer_size: usize) type {
                 const diff = @as(i128, seq) - @as(i128, pos);
 
-                if (diff == 0 and utils.tryCASAddOne(&self.enqueue_pos, pos, .monotonic) == null) {
+                if (diff == 0 and utils.tryCASAddOne(&self.enqueue_pos, pos, .release) == null) {
```
good catch, complete brainfart on my part to have gone with monotonic here

joadnacer commented on src/jdz_allocator.zig:
```diff
@@ -48,7 +48,7 @@ pub const JdzAllocConfig = struct {
     /// JdzSharedAllocator batch arena instantiation amount
     /// prevents allocator-induced false sharing if greater than total number of allocating threads
-    shared_arena_batch_size: u32 = 8,
+    shared_arena_batch_size: u32 = if (builtin.is_test) 2 else 8,
```
I don't think this should be set here, better to do it only via configs for tests which require (which you did for some tests already, no need to duplicate)

kython28 replied:
You are right

joadnacer commented on src/shared_allocator.zig:
```diff
@@ -659,3 +697,162 @@ test "consecutive overalignment" {
+    }
+    for (threads) |t| {
+        t.join();
+    }
```
multi-threaded tests are obviously non-deterministic, these are reasonable but I think should repeat it some reasonable number of times (a few hundred times at least probably, check compilation times)

joadnacer commented on src/shared_arena_handler.zig:
```diff
@@ -5,27 +5,71 @@ const utils = @import("utils.zig");
 
+
         pub fn init() Self {
+            const slot: usize = @atomicRmw(usize, &slots_counter, .Add, 1, .acquire);
```
this should probably use monotonic ordering?

kython28 replied:
I don't think so, i need to know what happened before to get a slot. I tested with monotonic before and it didn't work haha

joadnacer commented on src/shared_arena_handler.zig:
```diff
@@ -5,27 +5,71 @@ const utils = @import("utils.zig");
+                // Maximum number of SharedArenaHandler instances has been reached.
+                // This implementation limits the total number of concurrent 
+                // JdzAllocator instances to MAX_SLOTS (256).
+                @panic("Maximum number of SharedArenaHandler instances exceeded");
```
this is not ideal, it's somewhat of a stupid use-case but should probably try to support arbitrary number of allocator instances

kython28 replied:
Yes, you are right. Although I don't know how to do it, hahaha. Any idea would be nice. I think it would not be a problem because normally you always create one instance at the start of the program.

We can remove it, but it means we will need to remove the thread-local variable where I cache the thread arena

joadnacer commented:
currently on holiday until next sunday so not able to test anything locally, left some comments from reading the pr

most of the PR is sensible with just a few nitpicks, the `shared_arena_handler` changes seem potentially like a good direction to take but I need to think more about them, and do some testing. Would also like to see how it looks on [mimalloc-bench](https://github.com/daanx/mimalloc-bench) since the benchmarks in this repo are not great - iirc you'll just need to edit [this file](https://github.com/daanx/mimalloc-bench/blob/b517ae367c186be7365d238806b475478637d8f2/bench.sh) to run libjdzglobalwrap and libjdzshared, altough I can also just test it once home

kython28 committed:
chore: Bump version to 0.2.2 and improve test stability with multiple iterations

kython28 commented:
Thank you, man. I have also more changes ready, hahaha. I was waiting for your review and merge of this PR to create the new one.

https://github.com/kython28/jdz_allocator/compare/0.2.0...0.3.1

kython28 committed:
chore: Bump version to 0.2.3 and update arena initialization with @splat

kython28 committed:
refactor: Simplify arena initialization using @splat for spans and free lists

joadnacer commented:
able to get to this to now

seems building of libjdzshared/libjdzglobal is broken still, would be good to fix these - libjdzglobal is trivial 0.14.0 update, libjdzshared is not as the shared allocator can no longer be built at comptime (which would be nice to keep) - running `zig build` will try and compile these two files
can probably be solved via use of some branching on @ifComptime in the shared arena handler init
also grow_shrink_bench doesn't compile (just need to rename std.rand -> std.random)

hacked some fixes to the above to be able to run on mimalloc-bench and got following results (kjdz = your PR version, jdz = current main), huge improvements on shared allocator (jdzl) with your changes, especially on multi-threaded workloads (with the exception of rptest, but for some reason this benchmark tends to have odd results, best performing allocators here usually do badly on all other benchmarks) - perf is still a bit slower than for global allocators but close enough that I'll no longer feel bad about the shared alloc perf

```
#------------------------------------------------------------------
# test    alloc   time  rss    user  sys  page-faults page-reclaims
cfrac       jdz   03.45 3788 3.45 0.00 1 380
cfrac       kjdz  03.51 3692 3.51 0.00 9 380
cfrac       jdzw  03.45 3532 3.45 0.00 0 380
cfrac       kjdzw 03.50 3604 3.49 0.00 9 381
cfrac       jdzl  04.39 3932 4.39 0.00 0 603
cfrac       kjdzl 03.71 3168 3.70 0.00 9 382
espresso    jdz   03.67 6308 3.66 0.00 2 1122
espresso    kjdz  03.65 6360 3.65 0.00 0 1117
espresso    jdzw  03.67 6140 3.66 0.01 0 1118
espresso    kjdzw 03.69 6280 3.68 0.01 0 1120
espresso    jdzl  04.00 6708 3.99 0.00 0 1342
espresso    kjdzl 03.74 5948 3.74 0.00 0 1131
barnes      jdz   01.81 59076 1.78 0.03 1 16588
barnes      kjdz  01.82 59100 1.81 0.01 0 16590
barnes      jdzw  01.81 59012 1.80 0.01 0 16589
barnes      kjdzw 01.81 59172 1.80 0.01 0 16589
barnes      jdzl  01.81 59472 1.78 0.02 0 16812
barnes      kjdzl 01.80 58724 1.78 0.02 0 16598
redis       jdz   2.819 8996 0.12 0.03 13 1288
redis       kjdz  2.720 9116 0.12 0.02 0 1291
redis       jdzw  3.120 9004 0.14 0.02 0 1292
redis       kjdzw 2.819 9032 0.12 0.03 0 1291
redis       jdzl  3.660 9520 0.16 0.02 0 1513
redis       kjdzl 2.780 8700 0.14 0.00 0 1302
leanN       jdz   09.47 244620 22.30 0.35 87 58363
leanN       kjdz  09.42 249548 22.07 0.35 0 59444
leanN       jdzw  09.41 249020 21.99 0.42 0 59325
leanN       kjdzw 09.59 253576 22.45 0.34 0 60555
leanN       jdzl  10.43 254700 25.44 0.38 0 61040
leanN       kjdzl 09.94 890720 23.02 0.55 0 225690
larsonN-sized jdz   4.226 11189876 74.31 4.53 1 2886556
larsonN-sized kjdz  4.404 10728028 75.48 4.26 3 2767281
larsonN-sized jdzw  4.027 210268 79.33 0.21 5 52627
larsonN-sized kjdzw 4.011 223312 79.22 0.28 0 55762
larsonN-sized jdzl  17.390 186960 76.59 0.16 1 46233
larsonN-sized kjdzl 5.097 379704 79.05 0.28 0 95342
rptestN     jdz   2.331 154120 6.73 1.33 4 38810
rptestN     kjdz  2.286 157300 6.64 1.08 0 39673
rptestN     jdzw  2.366 154648 7.30 1.17 0 38964
rptestN     kjdzw 2.317 154660 6.68 1.18 0 38949
rptestN     jdzl  2.308 110052 6.74 1.19 0 27517
rptestN     kjdzl 4.245 759052 11.08 2.33 0 200629
gs          jdz   00.76 44564 0.69 0.03 146 7246
gs          kjdz  00.72 44692 0.70 0.02 0 7344
gs          jdzw  00.72 44476 0.71 0.01 0 7344
gs          kjdzw 00.72 44580 0.71 0.01 0 7344
gs          jdzl  00.72 44644 0.71 0.00 0 7558
gs          kjdzl 00.72 43976 0.70 0.01 0 7352
lua         jdz   04.03 101156 3.54 0.34 890 177138
lua         kjdz  03.89 101204 3.61 0.25 0 177549
lua         jdzw  03.90 101200 3.54 0.32 0 177693
lua         kjdzw 03.94 101164 3.59 0.32 0 177811
lua         jdzl  03.96 101564 3.67 0.25 0 200768
lua         kjdzl 03.91 100820 3.62 0.26 0 178809
alloc-test1 jdz   02.72 14416 2.72 0.00 1 2777
alloc-test1 kjdz  02.78 14484 2.77 0.00 0 2775
alloc-test1 jdzw  02.74 14284 2.74 0.00 0 2779
alloc-test1 kjdzw 02.75 14472 2.74 0.00 0 2776
alloc-test1 jdzl  02.96 14244 2.95 0.00 0 2897
alloc-test1 kjdzl 02.79 13728 2.78 0.00 0 2724
alloc-testN jdz   04.32 23152 68.71 0.00 0 5368
alloc-testN kjdz  04.41 23284 69.87 0.07 0 5364
alloc-testN jdzw  04.46 23132 70.68 0.01 0 5365
alloc-testN kjdzw 04.57 23516 72.40 0.02 0 5365
alloc-testN jdzl  13.70 21692 174.86 0.02 0 4961
alloc-testN kjdzl 05.65 28184 89.40 0.07 0 6612
sh6benchN   jdz   00.34 258792 4.43 0.13 0 65323
sh6benchN   kjdz  00.36 258988 4.84 0.11 0 65325
sh6benchN   jdzw  00.31 259248 4.53 0.15 0 65323
sh6benchN   kjdzw 00.35 258456 4.84 0.15 0 65323
sh6benchN   jdzl  02.92 253236 43.59 0.09 0 63398
sh6benchN   kjdzl 00.39 261200 5.35 0.14 0 66313
sh8benchN   jdz   00.88 143140 12.52 0.07 0 35607
sh8benchN   kjdz  00.87 174788 12.29 0.07 0 44194
sh8benchN   jdzw  00.85 149072 12.33 0.06 0 37093
sh8benchN   kjdzw 00.86 185796 12.45 0.09 0 47082
sh8benchN   jdzl  06.53 151220 98.49 0.07 0 37903
sh8benchN   kjdzl 01.02 432608 14.54 0.18 0 109230
xmalloc-testN jdz   0.473 647052 68.35 1.71 1 162762
xmalloc-testN kjdz  0.471 800660 68.40 1.89 0 201457
xmalloc-testN jdzw  0.501 474688 66.36 1.99 0 118576
xmalloc-testN kjdzw 0.476 785996 67.94 1.79 0 198697
xmalloc-testN jdzl  1.302 67948 74.15 0.59 0 17366
xmalloc-testN kjdzl 0.792 1610384 64.84 2.77 0 403586
cache-scratch1 jdz   00.87 4536 0.87 0.00 0 235
cache-scratch1 kjdz  00.88 4548 0.87 0.00 0 237
cache-scratch1 jdzw  00.87 4652 0.87 0.00 0 235
cache-scratch1 kjdzw 00.87 4636 0.86 0.00 0 237
cache-scratch1 jdzl  00.87 4824 0.87 0.00 0 460
cache-scratch1 kjdzl 00.87 4180 0.86 0.00 0 249
cache-scratchN jdz   00.12 4628 1.85 0.00 0 266
cache-scratchN kjdz  00.12 4636 1.87 0.00 0 265
cache-scratchN jdzw  00.12 4612 1.87 0.00 0 269
cache-scratchN kjdzw 00.12 4672 1.88 0.00 0 269
cache-scratchN jdzl  00.14 5008 1.79 0.00 0 493
cache-scratchN kjdzl 00.12 4252 1.90 0.00 0 280
glibc-simple jdz   01.80 3172 1.79 0.00 0 467
glibc-simple kjdz  01.85 3016 1.84 0.00 0 467
glibc-simple jdzw  01.85 3056 1.83 0.00 0 468
glibc-simple kjdzw 01.90 3300 1.89 0.00 0 466
glibc-simple jdzl  02.71 2884 2.70 0.00 0 500
glibc-simple kjdzl 02.15 2300 2.15 0.00 0 291
glibc-thread jdz   1.376 37452 31.85 0.00 0 9043
glibc-thread kjdz  1.414 37352 31.86 0.02 0 9041
glibc-thread jdzw  1.405 36860 31.77 0.00 0 9045
glibc-thread kjdzw 1.453 37324 31.79 0.01 0 9044
glibc-thread jdzl  7.247 34896 31.79 0.01 0 8707
glibc-thread kjdzl 1.824 41928 31.88 0.02 0 10379
rocksdb     jdz   03.38 95880 3.20 0.39 76 20951
rocksdb     kjdz  03.30 94740 3.15 0.36 0 20704
rocksdb     jdzw  03.31 94692 3.11 0.39 0 20702
rocksdb     kjdzw 03.27 94712 3.02 0.46 0 20704
rocksdb     jdzl  03.36 93652 3.16 0.41 0 20599
rocksdb     kjdzl 03.33 102164 3.23 0.33 0 22708
```

great job - I think definitely worth accepting the instance limitation for this, will dig in to the code a bit more over next few days to see if any other issues but should be mergeable soon once `zig build` builds successfully

joadnacer commented:
hi, this looks good, wanna handle the last few issues and we can merge?

kython28 commented:
I think i didn't get you hahaha. I think you are talking about the last review, I left you some comments. Could you answer them about what to do? And I will act accordingly 😁

joadnacer commented:
> seems building of libjdzshared/libjdzglobal is broken still, would be good to fix these - libjdzglobal is trivial 0.14.0 update, libjdzshared is not as the shared allocator can no longer be built at comptime (which would be nice to keep) - running zig build will try and compile these two files
can probably be solved via use of some branching on @ifComptime in the shared arena handler init
also grow_shrink_bench doesn't compile (just need to rename std.rand -> std.random)

was referring to this  - for the instance limit I think it's fine, the perf increase is worth the cap, so all good on those comments

if fixing libjdzshared nicely is too hard/impractical then can just skip this part for now as it's not too important and mostly for benchmarking/testing, and have tested it with some hacked fix already

kython28 committed:
Merge branch 'main' into shared-allocator-without-mutex

kython28 commented:
Ok perfect. Let me see how can i solve it hahaha. It looks kind of complicated :(

kython28 committed:
feat: enable comptime initializations

For comptime, the initialization number is limited to 1. I think, this
is a temporal limitation while i think in something better.

kython28 commented:
Well, i have just push a temporal solution for it. I need to think in a better way. For now, we will have this limitation for 1 initialization at comptime. Let me know what do you think, or if you see another way.

kython28 committed:
test: add comptime initialization test for JdzAllocator

kython28 committed:
refactor: Update atomic memory ordering from monotonic to acq_rel for thread safety
